src/baseAtomWithQuery.ts

```diff
@@ -63,92 +50,221 @@ export function baseAtomWithQuery<
+      TError,
+      TData,
+      TQueryData,
+      TQueryKey
+    >(client, defaultedOptions)
+    // Is it good approach to use client as WeakMap key? Would observer get ever garbage collected?
+    // Why not assign it to scoped variable?
+    // When whole atom is no longer accessible via reference, then observer should also get garbage collected.
     observerCache.set(client, newObserver)
 
     return newObserver
   })
-  if (process.env.NODE_ENV !== 'production') {
-    observerAtom.debugPrivate = true
-  }
-
-  const dataAtom = atom((get) => {
+  const queryResultGetterAtom = atom((get) => {
     const observer = get(observerAtom)
+    // When this atom has value other than null, it means that query result
+    // has been obtained either by optimisticResultAtom or query subscription defined in onMount below.
+    // Value is set to null when query is resetting.
+    const queryResultAtom = atom<QueryObserverResult<TData, TError> | null>(
+      null
+    )
+    // Subscribe to query result updates. Note that this happens only when query did not throw promise or error
+    // (therefore it has to have valid result)
+    queryResultAtom.onMount = (set) => {
+      // tanstack/react-query useBaseQuery is wrapping 'set' with
+      // https://tanstack.com/query/latest/docs/reference/notifyManager#notifymanagerbatchcalls
+      // not sure is it relevant here.
+      const unsubscribe = observer.subscribe(set)
+      // comment and line from tanstack/react-query useBaseQuery:
+      // Update result to make sure we did not miss any query updates
+      // between creating the observer and subscribing to it.
+      observer.updateResult()
 
-    const currentResult = observer.getCurrentResult()
-
-    const resultAtom = atom(currentResult)
-    if (process.env.NODE_ENV !== 'production') {
-      resultAtom.debugPrivate = true
+      return () => {
+        // Is there any case where it makes sense to reset value on queryAtom unmount?
+        // I think that it will get garbage collected due to WeakMap usage by jotai.
+        // Doing this could result in returning outdated state in returnAtom.
+        // set(null);
+        unsubscribe()
+      }
     }
 
-    resultAtom.onMount = (set) => {
-      const unsubscribe = observer.subscribe((state) => {
-        set(state)
-      })
-      return unsubscribe
+    return queryResultAtom
+  })
+  // Simple counter which allows to retrigger optimisticResultAtom calculation, after resetting atom.
+  const optimisticResultResetAtom = atom(0)
+  const queryKeyAtom = atom((get) =>
+    hashKey(get(defaultedOptionsAtom).queryKey)
+  )
+  // Reset queryResult each time when queryKey has changed.
+  const resetQueryResultOnQueryKeyChangeAtom = atom(
+    (get, { setSelf }) => {
+      get(queryKeyAtom)
+
+      Promise.resolve().then(setSelf)
+    },
+    (get, set) => {
+      set(get(queryResultGetterAtom), null)
+      set(optimisticResultResetAtom, get(optimisticResultResetAtom) + 1)
     }
+  )
```

Comment:
@kalijonn this is new comparing to my initial suggestion.

Reset `resultAtom` (and therefore trigger `optimisticResultAtom`) when observed `queryKey` changes.